src/Paprika.Tests/Store/AbandonedTests.cs

```diff
@@ -250,4 +243,61 @@ public async Task Reuse_in_grow_and_shrink()
             await block.Commit(CommitOptions.FlushDataAndRoot);
         }
     }
+
+    [Test]
+    [Category(Categories.LongRunning)]
+    public async Task Abandoned_chain_creation_with_overflow()
+    {
+        // Minimum iterations required to overflow the abandoned list (AbandonedList.MaxCount).
+        const int minIterations = 60;
+
+        // Minimum accounts required to overflow an abandoned page (AbandonedPage.Payload.MaxCount * 2).
+        const int numAccounts = 2000;
+
+        var keccaks = Initialize(numAccounts);
+
+        // Set big value.
+        var accountValue = new byte[2900];
+        new Random(17).NextBytes(accountValue);
+
+        using var db = PagedDb.NativeMemoryDb(140000 * Page.PageSize, HistoryDepth);
+
+        // Start read only batch to ensure that new pages are allocated instead of reusing
+        // the abandoned pages.
+        using (var read = db.BeginReadOnlyBatch())
+        {
+            for (var i = 0; i < minIterations; i++)
+            {
+                using var block = db.BeginNextBatch();
+                foreach (var keccak in keccaks)
+                {
+                    block.SetAccount(keccak, accountValue);
+                }
+
+                block.VerifyDbPagesOnCommit();
+                await block.Commit(CommitOptions.FlushDataAndRoot);
+            }
+        }
+
+        // Try reusing the abandoned pages.
+        for (var i = 0; i < minIterations; i++)
+        {
+            using var block = db.BeginNextBatch();
+            foreach (var keccak in keccaks)
+            {
+                block.SetAccount(keccak, accountValue);
+            }
+
+            block.VerifyDbPagesOnCommit();
```

Comment:
Hmmm, there's no actual assert here on the structure of Abandoned pages. Should we update visitor maybe and allow to walk through the list? Or maybe, externalize the list somehow from the root and allow to inspect it here to assert chaining?

Reply:
Need to think more about this one, I don't see any easy implementation without disrupting the existing code. We can assert on at least that abandoned list is full, but to fetch the exact abandoned pages and chaining, the complete Address structure has to be exposed either to Root or to the test directly.

Reply:
Missing visitor actually did catch the issue of non-reachability of pages with incorrect chain implementation, so it does provide some assertion.

Reply:
Ah, I see. Then I agree. If we're good from the reachability pov it provides some test for not dropping the tail.